cpp/ql/src/definitions.qll

```diff
@@ -197,14 +201,12 @@ Top definitionOf(Top e, string kind) {
     // exclude nested macro invocations, as they will overlap with
     // the top macro invocation.
     not exists(e.(MacroAccess).getParentInvocation()) and
-    // exclude results from template instantiations, as:
-    // (1) these dependencies will often be caused by a choice of
-    // template parameter, which is non-local to this part of code; and
-    // (2) overlapping results pointing to different locations will
-    // be very common.
-    // It's possible we could allow a subset of these dependencies
-    // in future, if we're careful to ensure the above don't apply.
-    not e.isFromTemplateInstantiation(_)
+    (
+      includeTemplateInstantiations = true
+      or
+      includeTemplateInstantiations = false and
+      not e.isFromTemplateInstantiation(_)
+    )
```

Comment:
I think we should move the `isFromTemplateInstantiation` check up to the caller for the sake of performance. I'll go into more detail over Zoom later today.